[WIP] VTX-MSP support

jesuslg123 opened the pull request:
## Support for VTX MSP

Implement VTX MSP protocol support so the video channel can be changed from Betaflight, ELRS Lua Script or any other system that supports VTX MPS protocol.

The plan is be able to change algo power and bandwidth in the future.

### Setup

1. Update the `MSP-OSD` package in both ends, Caddx vista/air unit and goggles V1/V2.
2. Load a VTX table data on Betaflight (this will happen automatically when develop is completed)
3. The feature is disabled by default, in order to enable it open the `MSP-OSD` package settings on the WTF OS configurator.

## The flow

![image](https://github.com/fpv-wtf/msp-osd/assets/7817064/8f045aad-94bc-46ef-96c4-1a822b2f9120)

1. VTX startup up
2. VTX setup the Betaflight VTX table and mark it as READY
3. VTX request every 200ms the VTX-MSP data
4. VTX read the channel from the MSP message
5. VTX send the data to the goggles via UDP protocol.
6. Goggles get the message
7. Goggles update the channel sent from the VTX

## Technical details

The  VTX-MSP data, including the channel is being pulled every 200ms in a loop. The channel data is added to a data frame sent to the goggles pretty frequently.

Once the data is on the goggles we check if the channel is different to the one from the previous message.

The FC will only send VTX message when the VTX has been marked as READY. 

What does it mean ready?
 - It has VTX table

## TODO
1. Sync channel changes from googles to BF. This probably is better to be handled on the VTX side. There is no other system doing this now. Probably it won't be done now.
2. Automatically set a valid VTX table data on BF.
3. Enable power change (need to find a method)
4. Enable bitrate mode change (need to find a method)

## Known issues

1. If channel is changed from the goggles, the sync with the MSP-VTX is lost. Right now this channel change is not being controlled.
12. You can setup any channel, even if is not supported, it will simply not do the channel change. This happen for index channels bigger than 8 in 25mbits or 3 in 50mbits.

j005u commented on jni/msp_displayport_mux.c:
```diff
@@ -74,6 +74,8 @@ static uint8_t serial_passthrough = 1;
 static uint8_t compress = 0;
 static uint8_t no_btfl_hd = 0;
 
+uint16_t fc_vtx_channel = -1;
```
an uint (unsigned int) can never be nagative. either use 0 as the indicator for not set or change it to an int so you can use negative values.

this might work because the -1 will still translate to some value, but this pattern can lead to unexpected behavior.

also, since we're doing c and being precise this should probably be a int8 rather than 16, we don't need more than 256 values.

jesuslg123 replied:
You are totally right, I started it as simply value to store the value and the lack of practice with C did the rest. I will move it to int8, it will always store small values as we don't have more than 8-9 channels.

j005u commented on jni/util/vtx_manager.h:
```diff
@@ -0,0 +1,312 @@
+typedef struct __gs_gui_config __gs_gui_config, *P__gs_gui_config;
+typedef struct __gs_gui_config gs_gui_config_t;
+
+struct __gs_gui_config {
```
we should probably not re-define __gs_gui_config as it's already defined in https://github.com/fpv-wtf/msp-osd/blob/f9d017bf6a39d07a1e96a33220cfbf168e03aebb/jni/rec/rec_shim.h#L5585.

Best would be to move the common parts to a common gs_gui_shim.c and include that.

jesuslg123 replied:
I was already about the double declaration of some times but some other types for some declarations have different types on both files, but I will revisit this to about the duplication as I also agree that we should not have it like that.

jesuslg123 replied:
Duplicated definition removed :)

j005u commented on jni/util/vtx_manager.c:
```diff
@@ -0,0 +1,73 @@
+        return;
+    }
+
+    setupManager();
```
we should not do the setup over and over again on each channel call. you could:

- add an `if (setChannelPilotOriginal != NULL) { return }` to the top of setupManager() so it'll only run once no matter how many times you call it
- call setupManager() from the overlay code during main() init once and only once (as bonus points you get to skip it if the setting is off)

j005u commented on jni/util/vtx_manager.c:
```diff
@@ -0,0 +1,73 @@
+}
+
+void changeChannel(int8_t channel) {
+    if(!get_boolean_config_value(VTX_MPS_CONFIG_KEY)) {
```
Additional sanity checks are not bad but in my opinion if we consider "separation of concerns" then this file should be as purely as shim to get a clean interface to the GUI functions as possible. Therefor it should not care about msp-osd config state and assume that if it's being called it should do it's job.

The check should be in the overlay itself.

jesuslg123 replied:
Agree, I will move the config check to the caller class place.

jesuslg123 committed:
Improve VTX manager init and config

jesuslg123 committed:
Add support for VTX MSP

Implement VTX MSP protocol support so the video channel can be changed
from Betaflight, ELRS Lua Script or any other system that uses VTX MPS
protocol.

The plan is be able to change algo power and badnwidth.

jesuslg123 committed:
Add configuration to enable/disable VTX MSP. Default: disabled

jesuslg123 committed:
Adding public channel selection

Need to refactor the .h - testing for now

jesuslg123 committed:
setting vtx wip

jesuslg123 committed:
Only load the VTX Manager if enable on settings

Also some minor warnings corrections

jesuslg123 committed:
Update to support latest NDK version

Based on @mmosca PR changes I have also enable latest NDK
support.

jesuslg123 committed:
Remove duplciated header redeclaration

Commit:
Experimental: Trying to make public chnl method available
